### PR TITLE
Add test for parallel hdf5 and AssertThrow

### PR DESCRIPTION
**Pull request by dangars**
`Assert` and `AssertThrow` hang with parallel HDF5. This test shows the behaviour.

This is related to #10737

### CODE REVIEW

File: tests/base/hdf5_07.cc [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+                           HDF5::File::FileAccessMode::create,
+                           mpi_communicator);
+
+      write_test<std::vector, double>(data_file,
+                                      dataset_dimensions,
+                                      mpi_communicator,
+                                      pcout);
+
+      AssertThrow(Utilities::MPI::this_mpi_process(mpi_communicator) == 0,
+                  ExcInternalError());
```
Comment (dangars) [inline tests/base/hdf5_07.cc:523]: If this is replaced by Assert, the program will hang as well.

Comment (bangerth) [reply tests/base/hdf5_07.cc:523]: This code is equivalent to the following:
```
{
  const bool b = (Utilities::MPI::this_mpi_process(mpi_communicator) == 0);
  AssertThrow (b, ExcInternalError());
}
```
Can you find out whether it also hangs if you replace `AssertThrow` by `Assert` here? In other words, whether the problem is in the call to `Utilities::MPI::this_mpi_process` or in the `AssertThrow`?

Separately, do you actually need all of the code above? Or could `write_test()` be drastically reduced to just one piece?

Comment (dangars) [reply tests/base/hdf5_07.cc:523]: I don't have access to a computer this week.

I'll give you a thorough answer next week :)

File: tests/base/hdf5_07.cc [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+      write_test<FullMatrix, double>(data_file,
+                                     dataset_dimensions,
+                                     mpi_communicator,
+                                     pcout);
```
Comment (dangars) [inline tests/base/hdf5_07.cc:170]: Interestingly the test hangs with the following code
```
write_test<FullMatrix, double>(data_file, dataset_dimensions, mpi_communicator, pcout);
```
But the test does not hang with the following code
```
write_test<std::vector, double>(data_file, dataset_dimensions, mpi_communicator, pcout);
```

**Comment by dangars** [comment]
How does `Assert` and `AssertThrow` work in parallel/MPI?

Let's say that only one process does not fulfill an `Assert` condition. Does this process notify the rest of the processes that they have to raise an exception?

**Comment by drwells** [comment]
Good question. If there is an uncaught exception and you initialized things with `MPI_InitFinalize` then we detect it in that classes destructor:
```cpp
          if (std::uncaught_exceptions() > 0)
            {
              // do not try to call MPI_Finalize to avoid a deadlock.
            }
          else
            {
              const int ierr = MPI_Finalize();
              (void)ierr;
              AssertNothrow(ierr == MPI_SUCCESS, dealii::ExcMPI(ierr));
            }
```
At that point `mpirun` knows that a process exited abnormally and it can deal with it (usually by killing the other processes).

On the other hand if we have an assertion failure instead of calling `abort` we call `MPI_Abort`.

**Comment by dangars** [comment]
Thanks @drwells !
Your explanation will help me to find the origin of the problem

**Comment by dangars** [comment]
I debug the program and I think that the problem is an interaction between exceptions and HDF5.

To debug the program I run the test using 3 MPI processes and attached each MPI process to an independent gdb.

I debug two versions of the test. The first version is the test that you can find in this PR.

In the second version I removed the HDF5 code.

**First version**
When I arrive to the lines below. Process 0 will continue normally and processes 1 and 2 execute `ExcInternalError()`

```
AssertThrow(Utilities::MPI::this_mpi_process(mpi_communicator) == 0,
ExcInternalError());
```

Then processes 1 and 2 arrive to the line 1322 of exceptions.h
```
case throw_on_exception:
  throw e;
```
When processes 1 and 2 throw the exception, they freeze instead of exiting the process. Therefore the whole program freezes.

**Second version**
Now I remove the HDF5 code and leave the AssertThrow.

Process 0 continues normally and processes 1 and 2 arrive to the line 1322 of exceptions.h
```
case throw_on_exception:
  throw e;
```
As expected, after the exception is thrown by processes 1 and 2, all the processes exit and the stacktrace is shown.


**Solution**
I see several solutions
- We can give a CMake compile option in which instead throwing an exception deal.II aborts the program and generates a core dump. A core dump can then be used to debug the program and get the stacktrace.
- We try to dig more in order to understand the problem better, although I'm not sure which direction to take.
- We document  the behavior in the HDF5 documentation of deal.II. Although this does not help to debug a program.

**Comment by drwells** [comment]
If I have understood this correctly this means that exceptions behave differently in HDF5 code than in non-HDF5 code. I think its worth understanding what's going on here before we move on to possible fixes. I'll take a look at this in depth at some point over the next few days.

**Comment by dangars** [comment]
Thanks @drwells !

Yes, exceptions interact in a weird way with parallel HDF5. I think that the problem is with HDF5, not with deal.II. I will contact the HDF5 developers once understand the origin of the problem.

I think that the problem arises when a HDF5 handler is opened and one process throws an exception and the other processes do not throw the exception.

I added a couple of tests. You can do a diff to see the differences between the tests.

If you debug the tests, it is helpful to add a breakpoint at the line 1322 of exceptions.h (function `issue_error_noreturn()`)

**hdf5_07**
- A parallel HDF5 handler is opened
- An exception is thrown **by two processes of the three processes**
- The program freezes

**hdf5_08**
- An exception is thrown by two processes of the three processes
- The program exits as expected

**hdf5_09**
- A parallel HDF5 handler is opened
- An exception is thrown **by all the processes**
- The program exits as expected

**Comment by drwells** [comment]
Sorry for the long delay - I finally looked into this and I have an answer.

If we do MPI IO through HDF5 then the following happens:

1. On the processor that calls the exception we try to close the parallel MPI file (`H5F_try_close`) which ultimately results in a barrier (`PMPI_barrier`).
2. On the processor that does not call the exception we call `H5VL_dataset_write` and end up in `PMPI_allreduce`.

Hence I think that exceptions can't work here - we have to abort.

**Comment by dangars** [comment]
@drwells Thanks a lot for the exhaustive analysis! That's very impressive! It is definitively not easy to debug MPI programs.

I can document this in the HDF5 interface of deal.II.

**Comment by drwells** [comment]
You're welcome. The trick is to let things hang in the debugger and then hit ^C to get stack traces - its a good way to debug hangs like this.

**Comment by drwells** [comment]
@dangars what do you want to do with this PR?

**Comment by dangars** [comment]
I think that we should close the PR. You found the problem!

I updated the documentation in #11050

**Pull request closed without merging**